zappa/zappa.py

```diff
@@ -869,107 +848,99 @@ def schedule_events(self, lambda_arn, lambda_name, events):
 
         """
-
-        client = self.boto_session.client('events')
-        lambda_client = self.boto_session.client('lambda')
-
         for event in events:
-
             function = event['function']
             expression = event['expression']
             name = event.get('name', function)
             description = event.get('description', function)
 
             self.delete_rule(name)
-
             #   - If 'cron' or 'rate' in expression, use ScheduleExpression
             #   - Else, use EventPattern
             #       - ex https://github.com/awslabs/aws-lambda-ddns-function
 
+            if not self.credentials_arn:
+                self.credentials_arn = self.create_iam_roles()
+
             if 'cron' in expression or 'rate' in expression:
-                response = client.put_rule(
+                rule_response = self.events_client.put_rule(
                     Name=name,
                     ScheduleExpression=expression,
                     State='ENABLED',
                     Description=description,
                     RoleArn=self.credentials_arn
                 )
             else:
-                response = client.put_rule(
+                rule_response = self.events_client.put_rule(
                     Name=name,
                     EventPattern=expression,
                     State='ENABLED',
                     Description=description,
                     RoleArn=self.credentials_arn
                 )
 
-            response = lambda_client.add_permission(
+            if 'RuleArn' in rule_response:
+                logger.debug('Rule created. ARN {}'.format(rule_response['RuleArn']))
+
+            logger.debug('Adding new permission to invoke Lambda function: {}'.format(lambda_name))
+            permission_response = self.lambda_client.add_permission(
                 FunctionName=lambda_name,
                 StatementId=''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(8)),
                 Action='lambda:InvokeFunction',
                 Principal='events.amazonaws.com',
-                SourceArn=response['RuleArn'],
+                SourceArn=rule_response['RuleArn'],
             )
 
+            if permission_response['ResponseMetadata']['HTTPStatusCode'] != 201:
+                print('Problem creating permission to invoke Lambda function')
+                return
+
             # Create the CloudWatch event ARN for this function.
-            target_arn = lambda_arn
-            inp = json.dumps({'detail': function})
-            target_response = client.put_targets(
+            target_response = self.events_client.put_targets(
                 Rule=name,
                 Targets=[
                     {
-                        'Id': function,
-                        'Arn': target_arn,
-                        'Input': inp,
-                    },
+                        'Id': 'Id' + ''.join(random.choice(string.digits) for _ in range(12)),
```

Comment:
I get that 12 random digits is pretty safe from conflict, but I'd feel better if it had even less potential problems - is there a limit to the id length or can we do the current unix timestamp + a random number/number+letter string?